nestedtensor/nested/masking.py

```diff
@@ -1,111 +1,184 @@
+        if (mask.numel() == 0) or (mask.numel() == 1 and mask == True):
+            return tensor
+
+        if mask.dim() == 1:
+            tensors = [tensor[i] if mask[i] else None for i in range(len(mask))]
+            tensors = list(filter(lambda x: x is not None, tensors))
+            return _merge(tensors, nested_dim)
+
+        if mask.dim() > 1:
+            tensors = [nt_from_tensor_mask(t, m, nested_dim, dt) for (t, m) in zip(tensor, mask)]
+            if not all(t.numel() == 0 for t in tensors):
+                tensors = list(filter(lambda x: x.numel() > 0, tensors))
+            return _merge(tensors, nested_dim)
+        else:
+            return None
     else:
-        sub_nested_dim = None if nested_dim is None else nested_dim - 1
-        tensors = [nested_tensor_from_tensor_mask(
-            t, m, sub_nested_dim) for (t, m) in zip(tensor, mask)]
-        if not all(t.numel() == 0 for t in tensors):
-            tensors = list(filter(lambda x: x.numel() > 0, tensors))
-        if len(tensors) == 0:
-            return tensor.narrow(0, 0, 0)
-        return _merge(tensors, nested_dim)
+        inner_tensors = []
+        if (mask.numel() == 0) or (mask.numel() == 1 and mask == True):
+            for i in range(len(tensor)):
+                inner_tensors.append(nt_from_tensor_mask(tensor[i], mask, nested_dim - 1, dt))
+        elif (mask.numel() == 1 and mask == False):
+            inner_tensors.append(None)
+        else:
+            inner_tensors = [nt_from_tensor_mask(t, m, nested_dim - 1, dt) for (t, m) in zip(tensor, mask)]
+
+        # Filtering out None values which were ignored by mask
+        inner_tensors = list(filter(lambda x: x is not None, inner_tensors))
+        return creation.nested_tensor(inner_tensors)
+
+# Get max size per each dimension from all the passed tensors.
+def get_max_size(obj, res=[1]):
+    if isinstance(obj, list) or isinstance(obj, tuple):
+        for o in obj:
+            res = get_max_size(o, res)
+
+    if isinstance(obj, nestedtensor.nested.nested.NestedTensor):
+        tres = get_max_size(obj.unbind())
+        while len(tres) > len(res):
+                res.append(0)
+
+        res = [max(i, j) for (i, j) in zip(res, tres)]
+
+    if isinstance(obj, torch.Tensor):
+        # scalar
+        if obj.dim() == 0 and obj.numel() == 1:
+            res = [1]
+        else:
+            while len(obj.size()) > len(res):
+                res.append(0)
+
+            res = [max(i, j) for (i, j) in zip(res, obj.size())]
+
+    return res
+
+
+def get_tensor_mask(nt, shape):
+    def pad_nt(nt, shape):
+        res_tensor = []
+        res_mask = []
+
+        if isinstance(nt, torch.Tensor):
+            if nt.numel() == 0:
+                raise RuntimeError("Empty tensors are not yet supported.")
 
+            # Dont pad in case of a scalar
+            if nt.dim() == 0:
+                return nt.item(), nt.item()
 
+            tensor = pad_tensor_to_shape(nt, shape).tolist()
```

Comment:
tolist() is a very expensive operation, since it turns the entire tensor into a nested list of python types.

Reply:
As a first version this is fine as long as it's correct, but this definitely means we consider this code to be changed very soon.

Reply:
it will be moved to C++ anyways